File: GameServer/Repository/AccountRepository.cs
using MySqlConnector;
using UnityServer.Models;
using UnityServer.Repository.Interface;

namespace UnityServer.Repository
{
    public class AccountRepository : IAccountRepository
    {
        private const string SqlIsLoginIdExists =
            "SELECT COUNT(*) FROM account WHERE user_id = @userId";
        private const string SqlCreateAccount =
            "INSERT INTO account (user_id, user_pw, nickname) VALUES (@userId, @userPw, @nickname); SELECT LAST_INSERT_ID()";
        private const string SqlValidateLogin =
            "SELECT id, nickname FROM account WHERE user_id = @userId AND user_pw = @userPw";
        private const string SqlIsNicknameExists =
            "SELECT COUNT(*) FROM account WHERE nickname = @nickname";

        private readonly string _connectionString;

        public AccountRepository(string connectionString)
        {
            _connectionString = connectionString;
        }

        public bool IsLoginIdExists(string loginId)
        {
            using MySqlConnection connection = new MySqlConnection(_connectionString);
            connection.Open();

            using MySqlCommand command = new MySqlCommand(SqlIsLoginIdExists, connection);
            command.Parameters.AddWithValue("@userId", loginId);

            return Convert.ToInt32(command.ExecuteScalar()) > 0;
        }

        public int CreateAccount(string loginId, string password, string nickname)
        {
            using MySqlConnection connection = new MySqlConnection(_connectionString);
            connection.Open();

            using MySqlCommand command = new MySqlCommand(SqlCreateAccount, connection);
            command.Parameters.AddWithValue("@userId", loginId);
            command.Parameters.AddWithValue("@userPw", password);
            command.Parameters.AddWithValue("@nickname", nickname);

            return Convert.ToInt32(command.ExecuteScalar());
        }

        public LoginResult? ValidateLogin(string loginId, string password)
        {
            using MySqlConnection connection = new MySqlConnection(_connectionString);
            connection.Open();

            using MySqlCommand command = new MySqlCommand(SqlValidateLogin, connection);
            command.Parameters.AddWithValue("@userId", loginId);
            command.Parameters.AddWithValue("@userPw", password);

            using var reader = command.ExecuteReader();
            if (!reader.Read())
                return null;

            return new LoginResult(
                reader.GetInt32("id"),
                reader.GetString("nickname")
            );
        }

        public bool IsNicknameExists(string nickname)
        {
            using MySqlConnection connection = new MySqlConnection(_connectionString);
            connection.Open();

            using MySqlCommand command = new MySqlCommand(SqlIsNicknameExists, connection);
            command.Parameters.AddWithValue("@nickname", nickname);

            return Convert.ToInt32(command.ExecuteScalar()) > 0;
        }
    }
}

File: NodeServer/src/repositories/accountRepository.js
const { pool } = require('../config');

async function isLoginIdExists(loginId) {
    const [rows] = await pool.execute(
        'SELECT COUNT(*) AS cnt FROM account WHERE user_id = ?',
        [loginId]
    );
    return rows[0].cnt > 0;
}

async function isNicknameExists(nickname) {
    const [rows] = await pool.execute(
        'SELECT COUNT(*) AS cnt FROM account WHERE nickname = ?',
        [nickname]
    );
    return rows[0].cnt > 0;
}

async function createAccount(loginId, password, nickname) {
    const connection = await pool.getConnection();
    try {
        const [insertResult] = await connection.execute(
            'INSERT INTO account (user_id, user_pw, nickname) VALUES (?, ?, ?)',
            [loginId, password, nickname]
        );

        const accountId = insertResult.insertId;

        await connection.execute(
            'INSERT INTO user (account_id, level, exp, hp, mp, map) VALUES (?, 1, 0, ?, ?, 1)',
            [accountId]
        );

        return accountId;
    } finally {
        connection.release();
    }
}

async function validateLogin(loginId, password) {
    const [rows] = await pool.execute(
        'SELECT id, nickname FROM account WHERE user_id = ? AND user_pw = ?',
        [loginId, password]
    );

    if (rows.length === 0) {
        return null;
    }

    return { accountId: rows[0].id, nickname: rows[0].nickname };
}

module.exports = { isLoginIdExists, isNicknameExists, createAccount, validateLogin };
